sagemaker/pipeline/pipeline_parameters.py

```diff
@@ -0,0 +1,764 @@
+    instance_config: InstanceConfig
+    task_config: TaskConfig
+    partition_config: PartitionConfig
+    training_config: TrainingConfig
+    inference_config: InferenceConfig
+    script_paths: ScriptPaths
+    step_cache_expiration: str
+    update: bool
+
+    def __eq__(self, other):
+        if not isinstance(other, PipelineArgs):
+            return False
+        return asdict(self) == asdict(other)
+
+    def __hash__(self):
+        # Convert the object to a dictionary
+        obj_dict = asdict(self)
+
+        # Convert the dictionary to a JSON string
+        json_str = json.dumps(obj_dict, sort_keys=True)
+
+        # Create a hash of the JSON string
+        return hash(json_str)
+
+    def get_hash_hex(self):
+        """Get unique string hash to differentiate between pipeline settings.
+
+        We use this hash as a unique identifier of execution parameters
+        in the intermediate output paths of the pipeline. This enables
+        caching steps when the input parameters are the same for the same
+        pipeline.
+
+        Returns
+        -------
+        str
+            A hexadecimal string representation of the object's hash.
+        """
+        # Convert the object to a dictionary
+        obj_dict = asdict(self)
+        # Remove non-functional keys from the dictionary,
+        # these do not affect the outcome of the execution
+        obj_dict.pop("step_cache_expiration")
+        obj_dict.pop("update")
+
+        # Convert the dictionary to a JSON string and create an md5 hash
+        json_str = json.dumps(obj_dict, sort_keys=True)
+        return hashlib.md5(json_str.encode()).hexdigest()
+
+    def __post_init__(self):
+        if not self.instance_config.train_on_cpu:
+            assert self.instance_config.gpu_instance_type, (
+                "GPU instance type must be specified if not training on CPU, "
+                f"got {self.instance_config.train_on_cpu=} "
+                f"{self.instance_config.gpu_instance_type=}"
+            )
+            assert self.aws_config.graphstorm_pytorch_cpu_image_url, (
+                "Must use provide GPU image when training on GPU. "
+                "use --graphstorm-pytorch-gpu-image-url"
+            )
+
+        # Ensure we provide a GConstruct/GSProcessing config file when running construction
+        if (
+            "gconstruct" in self.task_config.jobs_to_run
+            or "gsprocessing" in self.task_config.jobs_to_run
+        ):
+            assert self.graph_construction_config.config_filename, (
+                "Need to provide a GConstruct/GSProcessing config file "
+                "when running graph construction."
+            )
+
+        # TODO: When using GConstruct+DistPart/GBConvert (possible?) provide
+        # the correct partition input JSON filename?
+        # if "gconstruct" in self.task_config.jobs_to_run:
+        #     if "dist_part" in self.task_config.jobs_to_run or "gb_convert" in self.task_config.jobs_to_run:
```

Comment:
I think we do need this.

Reply:
If someone can not run gconstruct and partition with a single instance, he/she needs to use gsprocessing+distpart.

Reply:
Yes in that case they'd need to re-deploy their pipeline to add the necessary steps. I'll come up with a test that works here.